### PR TITLE
Setup pytest and create tests for existing functions

### LINKED ISSUE #38: [Task]: Setup pytest and create tests for existing functions
### Feature Description
Set up pytest for the project and create unit tests for existing functions in parser.py and main.py.

### Why Is This Needed?
- Ensure code quality and prevent regressions
- Validate parsing and text cleaning functionality
- Provide a foundation for future test-driven development

### Suggested Solutions
- Add pytest to requirements.txt
- Create tests/ directory with test files
- Write tests for parse_epub, parse_pdf, clean_extracted_text functions
- Test error handling and edge cases

### Additional Context
Technical Details: Use pytest fixtures for test data, mock file operations where needed. Complexity: Medium. Time Estimate: 2-3 hours.

### PR DESCRIPTION
**Pull request by isfopo**
## Summary

This PR sets up pytest for the project and creates comprehensive unit tests for existing functions, as requested in issue #38.

### Changes Made

- Added `pytest` to `requirements.txt` for test framework
- Created `tests/` directory with test suite structure
- Implemented `test_parser.py` with 12 tests covering:
  - `clean_extracted_text()`: whitespace normalization, page number removal, header/footer filtering
  - `parse_epub()`: file validation, error handling, successful parsing
  - `parse_pdf()`: file validation, text extraction, OCR detection
- Added `test_main.py` with tests for `_is_valid_book_file()` (GUI tests have issues with QApplication)
- Used proper mocking with `unittest.mock` and `pytest` fixtures
- All parser tests pass (12/12)

### Technical Details

- Tests use `@patch` decorators for external dependencies
- Mock objects simulate file system and library behaviors
- Error cases and edge conditions are thoroughly tested
- Test coverage includes validation, parsing, and text cleaning

### Testing

- Verified test discovery and execution with `pytest`
- Confirmed no linting errors with pyflakes
- All parser function tests pass
- GUI tests require special handling due to PyQt6 QApplication conflicts

Closes #38

### CODE REVIEW

File: parser.py [hunk 1]
```diff
@@ -6,6 +6,46 @@
 from pypdf import PdfReader
 
 
+def clean_extracted_text(text: str) -> str:
+    """Clean extracted text by normalizing encoding, removing headers/footers, and cleaning whitespace.
+
+    Args:
+        text: Raw extracted text.
+
```
Comment (isfopo) [inline parser.py:15]: Redundant `import re` inside `clean_extracted_text` function. The `re` module is already imported at the top of the file. Remove this line to avoid unnecessary imports.

File: tests/test_parser.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    @patch('ebooklib.epub.read_epub')
+    def test_successful_parsing(self, mock_read):
+        """Test successful EPUB parsing."""
+        # Mock book and items
+        mock_book = mock_read.return_value
+        mock_item = type('MockItem', (), {
+            'get_type': lambda self: 9,  # ITEM_DOCUMENT
+            'get_content': lambda self: b'<p>Hello world</p>'
+        })()
```
Comment (isfopo) [inline tests/test_parser.py:73]: Incorrect patch target. Should be `@patch('pypdf.PdfReader')` instead of `@patch('parser.PdfReader')` since `PdfReader` is imported from `pypdf`.

File: tests/test_parser.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        mock_book.get_items.return_value = [mock_item]
+
+        with patch('pathlib.Path.exists', return_value=True):
+            result = parse_epub("test.epub")
+            assert "Hello world" in result
+
+
+class TestParsePdf:
+    """Test parse_pdf function."""
+
```
Comment (isfopo) [inline tests/test_parser.py:83]: Same issue: incorrect patch target. Should be `@patch('pypdf.PdfReader')`.

File: tests/test_parser.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    def test_file_not_exists(self):
+        """Test error when file does not exist."""
+        with pytest.raises(ValueError, match="File does not exist"):
+            parse_pdf("nonexistent.pdf")
+
+    @patch('pypdf.PdfReader')
+    def test_invalid_pdf(self, mock_reader):
+        """Test error for invalid PDF."""
+        mock_reader.side_effect = Exception("Invalid PDF")
+        with patch('pathlib.Path.exists', return_value=True):
```
Comment (isfopo) [inline tests/test_parser.py:93]: Incorrect patch target. Should be `@patch('pypdf.PdfReader')`.

File: tests/test_parser.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
+"""Tests for parser module."""
+
+import pytest
+from pathlib import Path
+from unittest.mock import patch, mock_open
```
Comment (isfopo) [inline tests/test_parser.py]: `mock_open` is imported but not used in the tests. Remove the unused import.

**Review by isfopo — COMMENTED** [review]
Overall, this PR makes good progress in setting up pytest and adding comprehensive tests for the parser functions. The test structure is well-organized and covers important edge cases.

However, there are a few issues that need to be addressed:

1. Redundant import in `parser.py`
2. Incorrect patch decorators in PDF tests (should target `pypdf.PdfReader`)
3. Unused import in test file

Please fix these before merging.

File: tests/test_main.py
Comment (isfopo) [inline tests/test_main.py:1]: The parser tests are well-written and pass successfully. The mocking is appropriate and covers error cases, successful parsing, and text cleaning functionality.

However, the tests in `test_main.py` fail due to PyQt6 GUI initialization issues. The current approach of patching `QApplication` doesn't prevent the crash when instantiating `MainWindow`. Consider refactoring the tests to avoid instantiating GUI components, or use a more comprehensive mocking strategy (e.g., patch before imports or use pytest fixtures for GUI setup).

The text cleaning logic in `clean_extracted_text` looks solid, but the short line filtering might be overly aggressive for some texts. Consider making the length threshold configurable or reviewing edge cases.

Overall, great addition of test coverage. Once the GUI test issues are resolved, this should be ready for merge.

File: parser.py [hunk 6]
```diff
@@ -2,10 +2,50 @@
...
+    Returns:
+        Cleaned text.
+    """
+
+    # Split into paragraphs (double newlines)
+    paragraphs = re.split(r'\n\s*\n', text)
+
+    cleaned_paragraphs = []
+    for para in paragraphs:
+        # Clean within paragraph: normalize whitespace
```
Comment (isfopo) [inline parser.py:25]: The `clean_extracted_text` function is a good addition for preprocessing. The logic for removing page numbers and normalizing whitespace is appropriate. However, the condition for skipping short lines (`len(line) < 10 and not line[0].isupper()`) might filter out legitimate short sentences or proper nouns. Consider increasing the threshold or refining the logic to be less aggressive.

**Review by isfopo — COMMENTED** [review]
## Review Summary

**Overall Assessment:** This PR makes a strong contribution by adding comprehensive test coverage and integrating text cleaning functionality. The parser tests are excellent and demonstrate good testing practices.

**Strengths:**
- Comprehensive test suite with 12 passing parser tests covering various scenarios
- Proper use of mocking for external dependencies
- Clean integration of text cleaning into parsing functions
- Good test organization with class-based structure

**Areas for Improvement:**
- GUI tests in `test_main.py` currently fail due to PyQt6 initialization issues
- Text cleaning logic may be too aggressive in filtering short lines

**Recommendation:** The PR is ready for merge after addressing the GUI test failures. The text cleaning refinements are optional but would improve robustness.

File: tests/conftest.py
Comment (isfopo) [inline tests/conftest.py:1]: Excellent work addressing the review feedback! The QApplication fixture resolves the GUI test crashes, and the refined text cleaning threshold strikes a better balance. All tests now pass, and linting is clean.

**Review by isfopo — COMMENTED** [review]
## Review Summary

**Overall Assessment:** The review feedback has been effectively addressed. The PR now provides robust test coverage with proper GUI test setup and refined text processing logic.

**Strengths:**
- QApplication fixture properly handles PyQt6 initialization for GUI tests
- Comprehensive mocking in tests ensures reliable execution
- Text cleaning threshold adjustment improves filtering accuracy
- All 18 tests pass successfully
- Clean code with no linting issues

**Changes Verified:**
- GUI tests now run without crashes thanks to the conftest.py fixture
- Text cleaning logic refined from <10 to <5 character threshold
- Proper mocking added for file existence checks in main tests

**Recommendation:** This PR is now ready for merge. The test suite provides excellent coverage and the code quality is high.

Approved for merge.

File: tests/conftest.py
Comment (isfopo) [inline tests/conftest.py:1]: The PR remains in excellent shape with all previous feedback addressed. No additional issues found.

**Review by isfopo — COMMENTED** [review]
## Review Summary

**Overall Assessment:** This PR has been thoroughly reviewed and all feedback has been addressed. The code is production-ready.

**Strengths:**
- Complete test suite with 18 passing tests
- Proper PyQt6 handling for GUI tests
- Clean, well-documented code
- Follows Python best practices

**No Issues Found:**
- All linting passes
- Tests execute successfully
- Code quality is high

**Recommendation:** Ready for merge. The PR successfully implements pytest setup and comprehensive testing for the project.

Approved for merge.
